include/RED4ext/Handle.hpp

```diff
@@ -1,42 +1,321 @@
+        {
+            decwref_fn(this);
+        }
+    }
+
+    void incwref() const noexcept
+    {
+        if (refCount)
+        {
+            InterlockedIncrement(&refCount->weakRef);
+        }
+    }
+
+    bool incref_nz() noexcept
+    {
+        return refCount ? refCount->incref_nz() : false;
+    }
+
+    void move_construct_from(HandleBase&& rhs) noexcept
+    {
+        instance = rhs.instance;
+        refCount = rhs.refCount;
+        rhs.instance = nullptr;
+        rhs.refCount = nullptr;
+    }
+
+    void weakly_construct_from(const HandleBase& other) noexcept
+    {
+        instance = other.instance;
+        refCount = other.refCount;
+        incwref();
+    }
+
+    template <typename T>
+    friend class WeakHandle;
+
+    bool try_construct_from_weak(const WeakHandle<T>& other) noexcept
+    {
+        if (other.refCount && other.refCount->incref_nz())
+        {
+            instance = other.instance;
+            refCount = other.refCount;
+            return true;
+        }
+        return false;
+    }
+
+// letting this public for debugging purposes
+public:
+    T* instance = nullptr;
+    RefCnt* refCount = nullptr;
+};
+
+
+template<typename T>
+class Handle
+    : public HandleBase<T>
+{
+public:
+    constexpr Handle() noexcept = default;
+    constexpr Handle(nullptr_t) noexcept
+    {
+    }
+
+    ~Handle()
+    {
+        decref();
+    }
+
+    explicit Handle(T* ptr)
+    {
+        // todo: if T isn't IScriptable it hasn't enable_shared_from_this
+        //       but we may be able to provide a pointer to a fake empty object first
+        //       since the shared_from_this part seems to be skipped if *(uint64_t*)((char*)ptr + 0x16) == 0
+        //       although it seems the destructor calls instance's methods that could break
+
+        static REDfunc<Handle* (*)(Handle*, T*)> ctor_fn(Addresses::Handle_ctor_shared_from_this);
+        ctor_fn(this, ptr);
+    }
+
+    Handle(const Handle& other) noexcept
+    {
+        other.incref();
+        instance = other.instance;
+        refCount = other.refCount;
+    }
+
+    Handle(Handle&& rhs) noexcept
+    {
+        this->move_construct_from(std::move(rhs));
     }
 
-    inline T& operator*() const
+    Handle& operator=(const Handle& rhs) noexcept
+    {
+        Handle(rhs).swap(*this);
+        return *this;
+    }
+
+    Handle& operator=(Handle&& rhs) noexcept
+    {
+        Handle(std::move(rhs)).swap(*this);
+        return *this;
+    }
+
+    explicit Handle(const WeakHandle<T>& other)
+    {
+        if (!try_construct_from_weak(other))
+            throw std::bad_weak_ptr();
+    }
+
+    void swap(Handle& other) noexcept
+    {
+        do_swap(other);
+    }
+
+    void reset() noexcept
+    {
+        Handle().swap(*this);
+    }
+
+    void reset(T* ptr)
+    {
+        Handle(instance).swap(*this);
+    }
+
+    template<typename U = T, std::enable_if_t<!std::is_void_v<U>, int> = 0>
```

Comment:
Hmm, you can't use enable_if without template argument deduction... except if I missed something between C++14 and 17. I can't use T there because it would become and invalid non-type template parameter... The only alternative I see is `template<typename U = T, typename = std::enable_if_t<!std::is_void_v<U>>>`
U must stay.. to force deduction. This is more ugly, int = 0 is standard..
Other solution is `template<typename U = T> inline std::enable_if_t<!std::is_void_v<U>, T>* ` which is even uglier imo.

No need for typename with enable_if_t which is already a using decl I think.